File: Merge sort/mergesort.py

def merge(arrX,arrY,array):
    i=0
    while len(arrX) and len(arrY):
        x=arrX[0]
        y=arrY[0]
        if x<y:
            array[i]=x
            arrX.pop(0)
        else:
            array[i]=y
            arrY.pop(0)
        i+=1
    for j in arrX:
        array[i]=j
        i+=1
    for j in arrY:
        array[i]=j
        i+=1    
    return array

def mergeSort(array):
    length=len(array)
    if length<2:
        return array
    arrX=mergeSort(array[:length//2])
    arrY=mergeSort(array[length//2:])
    return merge(arrX,arrY,array)


print(mergeSort([8,7,6,5,4,3,12]))
#print(merge([1,2,8],[1,4,6],[9,10,11,12,13,14]))

File: Merge sort/mergesort.cpp
#include <iostream>
using namespace std;

void merge( int arr[],int start,int end,int mid){
    int temp[end-start+1];
    int i=start;
    int j=mid+1;
    int k=0;

    while (i <= mid && j <= end){
        if (arr[i] < arr[j]){
            temp[k++]=arr[i++];
        }
        else{
            temp[k++] = arr[j++];
        }

    }
    while (i <= mid){
       temp[k++] = arr[i++];
    }
    while (j <= end){
       temp[k++] = arr[j++];
    }
    i=start;
    for (auto j:temp){
       arr[i++]=j;

    }
}

 void mergeSort(int arr[],int start,int end) {
     if (start < end) {
         int mid = (start + end) / 2;
         mergeSort(arr, start, mid);
         mergeSort(arr, mid + 1, end);
         merge(arr, start, end, mid);
     }
 }

int main(){


    int arr[]={70,9,500,66,99};

    mergeSort(arr,0,4);




    for (int i=0;i<sizeof(arr)/sizeof(int);i++){

        cout<<arr[i]<<" ";
    }

    return 0;

}